### ISSUE TITLE
Improve language options

### ISSUE THREAD

**Issue by ZerooCool** [PrivateBin/PrivateBin#642]
Two elements are not translated if I choose to display the language bar.
I chose to display French by default.

If I copy the link from my privatebin to google, the description displayed will be in English, the existing description by default.

If I change the language from the interface, for German for example, my notice that I wrote in French, in the configuration file, will always be in French.

Please rethink this mechanism;)

**Comment by ZerooCool** [comment]
Can you upgrade this 2 line in the i18n/fr.json file ?

at the end : 

```
    "Encrypted note on PrivateBin":
        "Encrypted note on PrivateBin",
    "Visit this link to see the note. Giving the URL to anyone allows them to access the note, too.":
        "Visit this link to see the note. Giving the URL to anyone allows them to access the note, too."
```

by

```
    "Encrypted note on PrivateBin":
        "Message chiffré sur PrivateBin",
    "Visit this link to see the note. Giving the URL to anyone allows them to access the note, too.":
        "Visiter ce lien pour voir la note. Donner l'URL à une autre personne lui permet également d'accéder à la note."
```

**Comment by ZerooCool** [comment]
To obtain the description of the site in French, I had to force the language into French, and, I authorized the change of language.

```
languageselection = true
languagedefault = "fr"
```

Despite everything, it seems to me that when the language is automatically detected, in my case, in French, if I copy paste the URL address in Discord, Facebook, ..., the description offered will be in English.

```
languageselection = false
;languagedefault = "fr"
```

Please check this point.

**Comment by elrido** [comment]
Thank you for contributing the updates.

Regarding the settings, please have a look at the [documentation of these two](https://github.com/PrivateBin/PrivateBin/wiki/Configuration#languageselection--languagedefault).

As soon as you enable the language selection drop down, it will set a cookie to preserve the language setting. This will initially be based on your browser language hints, but change if you select another one from the drop down.

The `languagedefault` only applies if your browsers language setting doesn't match with any of the offered translation. I think almost all browsers do include English in that setting as a fallback, so it will be unlikely to get used. If you do provide a languagedefault and have the selection disabled, then your page will only ever show up in that language and other translations get disabled.

**Comment by ZerooCool** [comment]
You did not understand the problem, I re-express myself (with a translator).

When the language is automatic, this is the default configuration, and, whether I publish my instance link on Facebook or Discord, the text displayed is not in French but in English.

However, I am French, I use a browser in French, and, I post on a facebook or discord configured in French, and, my ip is French.
Thus, I find that it is not normal that the description published is in English.

I then noted:
1- that the translation into French had not been carried out, for the 2 lines concerned.
2- that despite everything, the description always seems to be displayed in English! It's a problem ! HERE !

3- I had to force the default language, French, and only then, the description, following a sharing, is displayed in French.
Only, the automatic language system will no longer work, and, if I am Chinese, and, if I want to share the site with my Chinese friends, the description will then be forced into French.

**Comment by r4sas** [comment]
Don't forget that previews generated on Facebook/Discord/Twitter side, not on your PC. They didn't know about your language in browser, but not their crawler, which requests info about page.
So, I think, here we can't do something. You must contact them, not us.

**Comment by r4sas** [comment]
In that case we need tag `og:locale:alternate`.

**Comment by ZerooCool** [comment]
if I understood correctly, I replace in
```
./tpl/page.php
./tpl/bootstrap.php
```

`<html lang="<?php echo I18n::_('en'); ?>">`
 by
`<!--<html prefix="og: http://ogp.me/ns#">-->`

But, I tested on Discord, I don't have the impression that it works, I post in English.
Snapchat too,
telegram too

On FB I post in French, but, I suppose that the FB cache has a lot to do with it.
Same Twitter, in french, but, i suspect the cache.

So, i'm not sur, and, the cache disrupts the tests.

**Comment by ZerooCool** [comment]
One thing is sure ! It is that with discord, the message posted is well posted in English!

Warning ! I made a correction on the 2 proposed language strings, the last 2 strings of the fr language, were always in English! This may be the case for all languages!

It would therefore be necessary, among other things, to ask multilingual correspondents to participate in the tests, as well as to translate the last two strings of the file from their language.

Otherwise, in addition to the cache, there is a risk of having false positives, because of the language file.

 #650

**Comment by elrido** [comment]
I don't think we can solve this on our side, as the bots of these services can only retrieve one version of the open graph and AFAIK there is no standard way to send more then one language in a single HTML document. So we will always have to choose one language and then stick to it.

If you wan't your instance to use French only (a legitimate use case) you can totally do that - it is explained in the configuration document I sent you above. It explains all four use cases that we cover, yours is one of them.

**Comment by ZerooCool** [comment]
Sorry but I did not understand.
First you tell me we can't ...
After that my case is taken into account ...

I remain on the findings in use, I do not know how to correct, but, if I am in automatic for the language, it seems that Discord chooses to display English, and, on the other hand, Facebook displays in French.

If I understood correctly, the problem could come from Discord?

If you understand better, thank you for developing, proposing, looking for, because, I do not understand this aspect or the possibilities to correct it.

**Comment by elrido** [comment]
Sorry, we may be talking about two different use cases:

Regarding _language selection by human users_: That is what the `languageselection = true` does - it lets them choose a language in a drop-down and persists the selection via a cookie. I assumed that was what you talked about when I first replied.

Regarding _what language discord sees_: These services robot user agent will probably not send any language header to your privatebin instance, so they get to see the default, which is probably English. Keep in mind that discord will retrieve the site only once and uses that cached response for every user, regardless of what language setting or preference that user has on discord. So they totally can localize their own content for you, but probably don't forward your localization to privatebin. This is an assumption of mine, based on the behavior of other such services - I don't use discord and don't know their applications source code.

So in my second answer, I assume that you want to always present your instance in French, and in order to do that you do have to set French as the `languagedefault` BUT keep `languageselection = false`. This robs the users of the option to select a language, but should solve the problem that discord gets sent the English version of the og headers - everyone should get it in French in that scenario.

In total, you can choose between 4 different behaviours with these two option.

**Comment by ZerooCool** [comment]
So, if I understand correctly, Facebook can identify the language, because if I post the site on FB, it will manage to identify the site in French, even if English is the default.
Unless actually, it is a false positive because of the cache.
Anyway, discord actually does not work the same way, I do not know more.

In the meantime, I set to true for the choice of language and I forced to default fr.
I would test again, later, during a next installation, on a new server and new domain.

Thank you for your explanation, it's very clear.
